checkpoint/recovery.go

```diff
@@ -176,15 +178,50 @@ func recoverFromLocalFile(
 		log.Int("num accounts", len(data.accounts)),
 		log.Int("num atxs", len(data.atxs)),
 	)
-	deps, proofs, err := collectOwnAtxDeps(logger, db, localDB, cfg, data)
-	if err != nil {
-		logger.With().Error("failed to collect deps for own atx", log.Err(err))
-		// continue to recover from checkpoint despite failure to preserve own atx
-	} else if len(deps) > 0 {
-		logger.With().Info("collected own atx deps",
-			log.Context(ctx),
-			log.Int("own atx deps", len(deps)),
-		)
+	allDeps := make([]*types.VerifiedActivationTx, 0)
+	allProofs := make([]*types.PoetProofMessage, 0)
+	if cfg.PreserveOwnAtx {
+		for _, nodeID := range cfg.NodeIDs {
+			deps, proofs, err := collectOwnAtxDeps(logger, db, localDB, nodeID, cfg.GoldenAtx, data)
+			if err != nil {
+				logger.With().Error("failed to collect deps for own atx",
+					nodeID,
+					log.Err(err),
+				)
+				// continue to recover from checkpoint despite failure to preserve own atx
+				continue
+			}
+
+			logger.With().Info("collected own atx deps",
+				log.Context(ctx),
+				nodeID,
+				log.Int("own atx deps", len(deps)),
+			)
+			allDeps = append(allDeps, deps...)
+			allProofs = append(allProofs, proofs...)
+
+			// deduplicate allDeps and allProofs by sorting and compacting
+			// then sort them by publishEpoch
+			slices.SortFunc(allDeps, func(i, j *types.VerifiedActivationTx) int {
+				return bytes.Compare(i.ID().Bytes(), j.ID().Bytes())
+			})
+			allDeps = slices.CompactFunc(allDeps, func(i, j *types.VerifiedActivationTx) bool {
+				return i.ID() == j.ID()
+			})
+			slices.SortFunc(allDeps, func(i, j *types.VerifiedActivationTx) int {
+				return int(i.PublishEpoch) - int(j.PublishEpoch)
+			})
+			slices.SortFunc(allProofs, func(i, j *types.PoetProofMessage) int {
+				iRef, _ := i.Ref()
+				jRef, _ := j.Ref()
+				return bytes.Compare(iRef[:], jRef[:])
+			})
+			allProofs = slices.CompactFunc(allProofs, func(i, j *types.PoetProofMessage) bool {
+				iRef, _ := i.Ref()
+				jRef, _ := j.Ref()
+				return iRef == jRef
+			})
```

Comment:
The problem is that entries returned from `maps.Values` are randomly sorted, but to apply during restore successfully they have to be sorted by `PublishEpoch`. Additionally I want to sort them by ID, so that two recovery files are easily comparable. Wdyt?

Reply:
Why not sort the slice returned from `maps.Valules()`?

Reply:
That would be more code that also uses more memory (2 additional maps) to achieve the same effect:

```go
			// deduplicate allDeps and allProofs by sorting and compacting
			// then sort them by publishEpoch
			mapDeps := make(map[types.ATXID]*types.VerifiedActivationTx, len(allDeps))
			for _, dep := range allDeps {
				mapDeps[dep.ID()] = dep
			}
			allDeps = maps.Values(mapDeps)
			slices.SortFunc(allDeps, func(i, j *types.VerifiedActivationTx) int {
				return bytes.Compare(i.ID().Bytes(), j.ID().Bytes())
			})
			slices.SortStableFunc(allDeps, func(i, j *types.VerifiedActivationTx) int {
				return int(i.PublishEpoch) - int(j.PublishEpoch)
			})
			mapProofs := make(map[types.PoetProofRef]*types.PoetProofMessage, len(allProofs))
			for _, proof := range allProofs {
				ref, _ := proof.Ref()
				mapProofs[ref] = proof
			}
			allProofs = maps.Values(mapProofs)
			slices.SortFunc(allProofs, func(i, j *types.PoetProofMessage) int {
				iRef, _ := i.Ref()
				jRef, _ := j.Ref()
				return bytes.Compare(iRef[:], jRef[:])
			})
```

It isn't even faster, because `slices.CompactFunc` only iterates once over the slice, which we have to do anyway to create the map?

Reply:
>That would be more code

It doesn't need to be. All `collect*()` functions could be refactored to return maps and you can merge 2 maps with `maps.Copy(dst, src)`

>also uses more memory (2 additional maps)
>It isn't even faster

Not necessarily, especially if there are many duplicates.

More important for me is understandable code. Maps are known to not allow duplicates, while I had to look the compact up and analyze the code to understand how it works (i.e. why it is important to sort before compacting :wink:). Anyway, I'm not insisting.